### PR TITLE
DDF-4689 Add backend support for sorting searches

### PR DESCRIPTION
**Pull request by bellcc**
#### What does this PR do?
- Provides the ability to order searches based on an attribute and sort order. 

#### Who is reviewing it? 

@GabrielFabian 
@cantstoptheunk 

#### Select relevant component teams: 
@codice/data 

#### Ask 2 committers to review/merge the PR and tag them here.
@andrewkfiedler
@ethantmanns

#### How should this be tested?

- Test the retrieve all endpoint against default values for `attr` and `sort_by`
- Test the retrieve all endpoint for other attributes. 
- Test the retrieve all endpoint for `asc` and `desc`

#### Any background context you want to provide?
- None

#### What are the relevant tickets?
For Jira:
[](https://codice.atlassian.net/browse/)

For GH Issues:
Fixes: #4689 

#### Screenshots

None

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by ethantmanns** [comment]
This is really about sorting, right? I'm confused as to why we're referring to it as filtering

**Comment by bellcc** [comment]
@ethantmanns Filtering may not have been the best way to describe this change. It is now `sorting`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java [hunk 1]
```diff
@@ -64,6 +64,12 @@
 
   private static final String COUNT = "count";
 
+  private static final String SORT_BY = "sort_by";
+
+  private static final String ATTR = "attr";
+
+  private static final String ASCENDING = "asc";
```
Comment (ethantmanns) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: ❓ Is there anywhere we can share constants for these so it's easier to maintain?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java [hunk 2]
```diff
@@ -183,7 +197,7 @@ String getSubjectIdentifier() {
     return subjectIdentity.getUniqueIdentifier(SecurityUtils.getSubject());
   }
 
-  private int getOrDefaultParam(Request request, String key, int defaultValue) {
+  private static int getOrDefaultParam(Request request, String key, int defaultValue) {
```
Comment (ethantmanns) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: ❓ Just curious why the change to static here, why not the function below as well?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java [hunk 3]
```diff
@@ -98,6 +104,14 @@ public void init() {
           int start = getOrDefaultParam(req, START, MIN_START);
           int count = getOrDefaultParam(req, COUNT, MAX_PAGE_SIZE);
 
+          String attr = req.queryParams(ATTR);
+
+          if (attr == null) {
```
Comment (ethantmanns) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: ❓ Not sure how this works, do we also need to use default if empty or invalid attribute? Might go out of scope for that second one.

Comment (bellcc) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: As it currently exists this endpoint returns all searches in descending order based on their last modified attribute. 

If there are preferred defaults we will find them as we iterate. 

Modified the code to handle blank and invalid entries

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java (hunk 2 above)
Comment (bellcc) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: Fair point. I've made the other one `static` as well.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java (hunk 1 above)
Comment (bellcc) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/QueryMetacardApplication.java]: I'm not familiar with such a place where we're sharing constants like this for our endpoints.

**Comment by bellcc** [comment]
build now

File: distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+The queries endpoint can be used for one or more of these operations on an instance of ${branding}:
+
+* Create query metacards and store them in the ${branding} catalog.
+* Retrieve all query metacards stored in the ${branding} catalog and sort them based on attribute and sort order.
+* Retrieve a specific query metacard stored in the ${branding} catalog.
+* Update query metacards that are stored in the ${branding} catalog.
+* Delete query metacards that are stored in the ${branding} catalog.
+
+.Queries Endpoint URL
+[source,http]
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc]: This should probably be https. It won't affect the syntax highlighting, but there is a hover-over label that'll read HTTPS

File: distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc [hunk 5]
```diff
@@ -0,0 +1,241 @@
+:title: Queries Endpoint
+:type: endpoint
+:status: published
+:operations: crud
+:link: _queries_endpoint
+:summary: To perform CRUD (Create, Read, Update, Delete) operations on query metacards in the catalog, work with one of these endpoints.
+
+== {title}
+
+The queries endpoint enables an application to create, retrieve, update, and delete query metacards.
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc]: Should we define query metacards here? As to how they are different from other metacards?

File: distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+----
+https://<HOSTNAME>:<PORT>/search/catalog/internal/queries
+----
+
+=== Queries Endpoint Create Examples
+
+To create a query metacard through the queries endpoint, send a `POST` request to the queries endpoint.
+
+.Queries Endpoint Create Request Body
+[source,json]
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc]: can you make this `[source,json,linenums]`? It's nice to have the line numbers on samples longer than two or three lines

File: distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+	],
+	"type":"advanced",
+	"title":"Search Title"
+}
+----
+
+A successful create request will return a status of `201 CREATED`.
+
+.Queries Endpoint Create Success Response Body
+[source,json]
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc]: linenums here too

File: distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+|===
+
+A successful retrieval request will return a status of `200 OK`.
+
+=== Queries Endpoint Retrieve Examples
+
+.Retrieve Specific Query Endpoint URL
+[source,http]
+----
+https://<HOSTNAME>:<PORT>/search/catalog/internal/queries/:id
```
Comment (ricklarsen) [inline distribution/docs/src/main/resources/content/_integrating/_endpoints/query-endpoints.adoc]: We should offset the placeholder `id` with angle brackets, like `<HOSTNAME>` and `<PORT>` 
and, I suppose, do all caps, although I don't think I've been very consistent on the other endpoints.

**Comment by bellcc** [comment]
build now

**Comment by bellcc** [comment]
build now

**Comment by cantstoptheunk** [comment]
Hero Successful: 
Tested sorting on Query Metacard Types on the following attribute types:
  -Boolean (Enterprise)
  -String (Title, id, modified)
Note that Query Metacards don't contain numeric fields thus was not part of the hero.
